### PR TITLE
chore: accept payload on workspace usage route

### PR DESCRIPTION
**Pull request by f0ssel**
This adds new behavior to the workspace usage endpoint to accept a payload containing an `AgentID` and `AppName`. This is so the CLI can start informing us of activity based on different "defined apps". This is currently behind an experiment flag `workspace-usage` which it will stay behind until we are reading to migrate from `workspace_agent_stats` table to `workspace_app_stats`.

### CODE REVIEW

File: codersdk/workspaces.go [hunk 1]
```diff
@@ -316,10 +316,31 @@ func (c *Client) PutExtendWorkspace(ctx context.Context, id uuid.UUID, req PutEx
...
+var AllowedAppNames = []UsageAppName{
+	UsageAppNameVscode,
+	UsageAppNameJetbrains,
+	UsageAppNameReconnectingPty,
+	UsageAppNameSSH,
+}
+
 // PostWorkspaceUsage marks the workspace as having been used recently.
-func (c *Client) PostWorkspaceUsage(ctx context.Context, id uuid.UUID) error {
+func (c *Client) PostWorkspaceUsage(ctx context.Context, id uuid.UUID, req PostWorkspaceUsageRequest) error {
```
Comment (johnstcn) [inline codersdk/workspaces.go:341]: note: this is technically a breaking change in the public codersdk API; we should either call it out or make this a separate method.

File: coderd/workspaces_test.go [hunk 2]
```diff
@@ -3371,3 +3371,96 @@ func TestWorkspaceFavoriteUnfavorite(t *testing.T) {
...
+			AgentID: workspace.LatestBuild.Resources[0].Agents[0].ID,
+			AppName: "reconnecting-pty",
+		})
+		require.NoError(t, err)
+		// ssh works
+		err = client.PostWorkspaceUsage(ctx, r.Workspace.ID, codersdk.PostWorkspaceUsageRequest{
+			AgentID: workspace.LatestBuild.Resources[0].Agents[0].ID,
+			AppName: "ssh",
+		})
+		require.NoError(t, err)
```
Comment (johnstcn) [inline coderd/workspaces_test.go:3464]: Can we also validate that `ReportAgentStats` gets called and bumps the workspace deadline?

File: cli/portforward.go [hunk 3]
```diff
@@ -137,7 +137,7 @@ func (r *RootCmd) portForward() *serpent.Command {
 				listeners[i] = l
 			}
 
-			stopUpdating := client.UpdateWorkspaceUsageContext(ctx, workspace.ID)
+			stopUpdating := client.UpdateWorkspaceUsageContext(ctx, workspace.ID, codersdk.PostWorkspaceUsageRequest{})
```
Comment (johnstcn) [inline cli/portforward.go:140]: non-blocking: We might want to expose this as an app named "port-forward" or similar?

File: codersdk/workspaces.go (hunk 1 above)
Comment (f0ssel) [reply codersdk/workspaces.go:341]: I'll do a separate method 👍

File: coderd/workspaces_test.go (hunk 2 above)
Comment (f0ssel) [reply coderd/workspaces_test.go:3488]: I've updated this test to check the deadline has been bumped.

**Review by johnstcn — APPROVED** [review]

File: cli/portforward.go (hunk 3 above)
Comment (johnstcn) [reply cli/portforward.go]: Let's leave it for now. I'd like to avoid any insights changes in the course of this refactor.

**Pull request merged**